src/Microsoft.AspNet.Razor/CodeGenerators/CSharpTagHelperCodeRenderer.cs

```diff
@@ -179,68 +177,45 @@ protected virtual string GenerateUniqueId()
                                                           _tagHelperContext.ExecutionContextAddMethodName,
                                                           tagHelperVariableName);
                 }
-
-                // Render all of the bound attribute values for the tag helper.
-                RenderBoundHTMLAttributes(
-                    chunk.Attributes,
-                    tagHelperDescriptor.TypeName,
-                    tagHelperVariableName,
-                    tagHelperDescriptor.Attributes,
-                    htmlAttributeValues);
             }
         }
 
-        private void RenderBoundHTMLAttributes(
-            IList<KeyValuePair<string, Chunk>> chunkAttributes,
-            string tagHelperTypeName,
-            string tagHelperVariableName,
-            IEnumerable<TagHelperAttributeDescriptor> attributeDescriptors,
-            Dictionary<string, string> htmlAttributeValues)
+        private void RenderTagHelperPropertyVerification(
+            IEnumerable<KeyValuePair<string, Chunk>> chunkAttributes,
+            IEnumerable<TagHelperDescriptor> tagHelperDescriptors)
         {
+            if (_designTimeMode)
+            {
+                // No need to render any dictionary verification in design time mode.
+                return;
+            }
+
             // Track dictionary properties we have confirmed are non-null.
             var confirmedDictionaries = new HashSet<string>(StringComparer.Ordinal);
 
-            // First attribute wins, even if there are duplicates.
-            var distinctAttributes = chunkAttributes.Distinct(KeyValuePairKeyComparer.Default);
-
-            // Go through the HTML attributes in source order, assigning to properties or indexers as we go.
-            foreach (var attributeKeyValuePair in distinctAttributes)
+            foreach (var tagHelperDescriptor in tagHelperDescriptors)
             {
-                var attributeName = attributeKeyValuePair.Key;
-                var attributeValueChunk = attributeKeyValuePair.Value;
-                if (attributeValueChunk == null)
+                foreach (var attributeDescriptor in tagHelperDescriptor.Attributes)
```

Comment:
This loop will not handle precedence of non-indexer properties over indexers. For example if `asp-aaa` and `asp-bbb` were used in an element associated with a tag helper that has non-indexers with those names, an unused dictionary would still be initialized.

Need to instead check
``` c#
chunkAttributes.Any(
    keyValuePair => tagHelperDescriptor.Attributes.Any(
        attribute => attribute.IsNameMatch(keyValuePair.Key) && attribute.IsIndexer))
```